### DISCUSSION TITLE
Best approach to add wifi support

### DISCUSSION THREAD

**Discussion by nhi-vanye** [flatcar/Flatcar#969] [Q&A] [answered]
Hi,

I'm looking at using flatcar to replace an ageing embedded linux (ubuntu), but some of our installed systems connect via wifi.

Any hints on which of the many wifi tools have been most successful on flatcar ? If I need to use a systemd-sysext I can (already using them to add tailscale support)

Ideally I'd like to be able to hard-code the SSID/password into the ignition file to bring it into the network automatically

**Comment by pothos** [comment, marked as answer]
With WiFi not being supported in the Flatcar kernel, I think you may have to recompile the kernel unless you are very lucky and could just compile missing modules in the devcontainer. As of now, a custom kernel is not easy to marry with the official Flatcar releases, it would be your own build. If that use case is coming up more often, we should check how much size increase the wifi stack gives when enabled in the kernel to have some data to decide whether this is something to add by default. As far as I know Fedora CoreOS should have WiFi support, so maybe give it a try.
For WiFi tooling, IWD works well with systemd-networkd.

**Reply by nhi-vanye** [reply]
Thanks for that level of detail, I hadn't thought the drivers would not be present so that saves me a lot of wasted time.

I think I'll look at generating some data on driver sizes  to allow a decision on including them to be made. I only have a couple of wifi chips I need to support, but we need something for the general case.

If the sizes were deemed too prohibitive for general distribution would it be viable for the team to build them but just distribute them as a tarball and leave it to OEMs/others to solve the install/update issue. I'm still getting the benefit of not having to build drivers and I need to be looking at handling updates for my other systemd-sysexts anyway...

**Reply by pothos** [reply]
There are probably three parts, first the generic kernel support, second loadable kernel modules that could be brought in via a sysext, third any required binary firmware which could also be brought in with the sysext. So the easy question would be the size of the generic wifi support in the kernel that would allow for the other two things to be added later (either precompiled or compiled on the machine with the Flatcar devcontainer see https://www.flatcar.org/docs/latest/reference/developer-guides/kernel-modules/).

**Reply by nhi-vanye** [reply]
I couldn't get the sdk to run to build full system, but I did some ballpark kernel compilations using the dev container of stable (kernel 5.15.89-flatcar) with the current `.config` and one that enabled WLAN with selected drivers.

I haven't tested this kernel/drivers yet.

With the current `.config`  I obtained a `dir-pkg` with 901 files and total size of boot and lib/modules/ of 777MB.

Enabling a limited set of wifi drivers (hopefully focussing on server type chipsets - ADMTek, Atheros/Qualcomm, Atmel, Broadcom, Intel, and the `dir-pkg` was 809MB with 934 files.

At the high level that increase in size was distributed across the drivers as

| Vendor  | `du -sh` of `lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/*` |
| -------- | ------------- |
| ADMTek | 208k |
| Atheros | 9.5M |
| Atmel | 208k |
| Broadcom | 2.3M |
| Intel | 9.2M |

With some further refinement we could reduce the number of Atheros and Intel chipsets supported and yet still be useful for selected edge hardware (I'm currently using a mix of NUC-like systems as embedded computers inside lab equipment)

The compressed kernel increased from `12154944` to `12182304` - an increase of approx 0.25%

What size are you focussing on - the overall ISO size ?

I'll leave it at this point to give people time to discuss any next steps, happy to contribute in any way. It's been a trip down memory lane as I haven't built a linux kernel for almost 20 years - nothing's changed.

[I've attached the .config patch for further refinement discussion]

[add-wifi.diff.txt](https://github.com/flatcar/Flatcar/files/10895833/add-wifi.diff.txt)

**Reply by pothos** [reply]
Thanks for the test! The modules weren't compressed?
The size we care about is mostly the unified kernel+initramfs binary that lands on /boot. The loadable modules and the firmware binaries would normally not be part of this since we can try to load them from /usr or even better some additional image file.

**Reply by nhi-vanye** [reply]
The modules were compressed...

```
flatcardevelopercontainer /usr/src/linux # find wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ -type f -ls
  2262889    224 -rw-r--r--   1 root     root       227584 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ar5523/ar5523.ko.xz
  2262894    276 -rw-r--r--   1 root     root       279076 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ath10k/ath10k_pci.ko.xz
  2262895    252 -rw-r--r--   1 root     root       257640 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ath10k/ath10k_sdio.ko.xz
  2262893   2484 -rw-r--r--   1 root     root      2542088 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ath10k/ath10k_core.ko.xz
  2262897   1564 -rw-r--r--   1 root     root      1601484 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ath5k/ath5k.ko.xz
  2262901   1980 -rw-r--r--   1 root     root      2025208 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ath9k/ath9k_hw.ko.xz
  2262902     96 -rw-r--r--   1 root     root        96972 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ath9k/ath9k_pci_owl_loader.ko.xz
  2262900    328 -rw-r--r--   1 root     root       332116 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ath9k/ath9k_common.ko.xz
  2262899   1312 -rw-r--r--   1 root     root      1341952 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ath9k/ath9k.ko.xz
  2262890    468 -rw-r--r--   1 root     root       476012 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/ath.ko.xz
  2262904    624 -rw-r--r--   1 root     root       636384 Mar  6 07:30 wifi-limited/lib/modules/5.15.89-flatcar/kernel/drivers/net/wireless/ath/wcn36xx/wcn36xx.ko.xz
```

I'm unfamiliar with building the combined kernel+initramfs, I saw the build needed a `bootengine.cpio` but it didn't create it so I just touched the file to get the kernel (`make all`) to build - assuming that the sizes would be at least indicative as I wasn't expecting wifi drivers to be in the initrd. 


Looks like the uncompressed kernel size went from `515757488` to `517421712` about ~1.6MB increase.

```
flatcardevelopercontainer /usr/src/linux # ls -l current/boot/vmlinux-5.15.89-flatcar wifi-limited/boot/vmlinux-5.15.89-flatcar
-rwxr-xr-x 1 root root 515757488 Mar  5 23:20 current/boot/vmlinux-5.15.89-flatcar
-rwxr-xr-x 1 root root 517421712 Mar  6 07:36 wifi-limited/boot/vmlinux-5.15.89-flatcar
```

On my stable system, the current stable combined image is `57783968` bytes

```
core@unset ~ $ ls -l /boot/flatcar/vmlinuz-b
-rwxr-xr-x. 1 root root 57783968 Mar  3 16:30 /boot/flatcar/vmlinuz-b
```
